#2805 Upgrade Node

YuaFox opened the pull request:
Upgrade to node v22 LTS and update libraries

This is an experimental pull request

YuaFox committed:
#2805 Upgrade Node
- Nodejs upgraded to v22 LTS
- Cypress upgraded

Limraj commented:
Hi @YuaFox,
before committing you can rebuild locally using task: buildRunDebugProdTestUi;
If this task completed locally, it should also works here.

Remember that we have two vue projects, where each time you change the version you have to remove node_modules in places:
1. Delete ..\Scada-LTS\scadalts-ui\node_modules ;
2. Git->Rollback on ..\Scada-LTS\scadalts-ui (lib @min-gb in soruce)
3. Delete ..\Scada-LTS\WebContent\resources\node_modules
4. Run task buildRunDebugProdTestUi;

Regards,
Kamil Jarmusik

YuaFox committed:
#2805 Upgrade Node
- Added router to tests
- Updated @vue/test-utils
- Updated package-lock.json

Limraj commented on scadalts-ui/.gitignore:
```diff
@@ -29,5 +29,7 @@ test-results.xml
+!@min-gb/dist
+!@min-gb/public
+!@min-gb/src
+!@min-gb/*.*
```
We don't want it in ignore, this library is used in our code, and since it is no longer available in the remote repository we have moved it to sources, what lies ahead is to integrate this library in our sources:
https://github.com/SCADA-LTS/Scada-LTS/issues/2957

YuaFox replied:
I'm aware of that, thats why i put the !, to include it in the repo. I changed it to that to ignore the node_modules folder inside @min-gb

Limraj replied:
Hi @YuaFox, 'to include it in the repo'? 
I mean what did you do because I don't understand...

If you set ignore on this library, git will not track changes in it, we don't want anything in the sources to be ignored by git.

Limraj replied:
Okay, I understand, you made it so that it won't be ignored? Great, we need to audit this library, what is it, etc. 
Can't we rewrite the functions of this library ourselves.

Limraj replied:
From what I can see, this configuration does not block the visibility of changes in min-gb, so there is no need to change it here, unless I am missing something, I just did a test and changed the file: \node_modules\@min-gb\vuejs-components\babel. config.js and git sees this change.

Limraj replied:
However, this is interesting because in the old gitignore file we do not ignore node_modules from scadalts-ui and yet the content of node_modules is ignored... if so, there is no point in changing it, although it is worth considering what causes this, it is probably some pattern from gitignore file...

Limraj replied:
@YuaFox  " I changed it to that to ignore the node_modules folder inside @min-gb" What is the justification for this? Because I don't remember it causing any problems...

YuaFox replied:
When i ran the tests a node-module folder was created inside @min-gb git tracked these changes and it was annoying. If you cant reproduce this with node 22 maybe it was my bad and I ran something I shouldn't.

Limraj replied:
Show me on screenshot..

Limraj replied:
We get errors like this, but it looks like the new front is working...
[node_logs.txt](https://github.com/user-attachments/files/16797512/node_logs.txt)

Theoretically, entries regarding not supported for a given library are disturbing. Especially regarding the Node engine.
One might wonder whether this somehow affects the final compilation of the new fronted, and running e2e tests.
You could still check whether these e2e tests work on this version, that is, change some condition and check whether it returns an error.

From what I see, the change to gitignore is not necessary unless you show me something else.

Limraj replied:
Okay, i'am see:
![image](https://github.com/user-attachments/assets/23d74e14-af5a-46f5-9957-6126ad51a721)

Limraj replied:
With your gitignore corrections:
![image](https://github.com/user-attachments/assets/53495a3d-b1b4-48d4-9829-66680a66b2c9)

That's okay.

YuaFox replied:
A lot of libraries Scada-LTS is using reached EOL including Vue 2.7. Upgrading to Vue 3 and removing deprecated libraries for new ones removes all warnings.

About the "Unsupported engine" error: https://github.com/vuejs/vue-cli/issues/7424

I can do it, but it will take time.

Limraj replied:
@YuaFox, 
sure, getting rid of deprecated lib would be of great value and help to us, over time, these versions of libraries may disappear from remote repositories. However, these unsupported entries worry me. **If this doesn't cause problems in compiling and running e2e tests, we can end this Issue here**. I'm afraid that switching to vue 3 will involve additional work in the source code of our application, so apart from the dependencies themselves, there will be more work, but from what I know, there are big differences between these versions of vue, I don't know if vue 3 is compatible with vue 2. So this would be a big topic here.

For now you can do this:
https://github.com/SCADA-LTS/Scada-LTS/issues/2955

Regards,
Kamil Jarmusik

YuaFox replied:
Vue 3 is compatible with vue 2 components, but there are some things that change. For example, this.$router no longer works and you have to do const router = useRouter(). I'll open an issue later about this topic and check if its a big change. There's a migration tool too.

All unit tests are working (Mocha and Java JUnit) but cypress is not, but i think those test are outdated right? I'm getting 404 errors to urls like http://localhost:8080/Scada-LTS/modern_watch_list.shtm , this test fails on the current release too.

Limraj replied:
http://localhost:8080/Scada-LTS/modern_watch_list.shtm 
You can delete this case, we no longer have this page.
all tests cypress not working?

YuaFox replied:
Some are failing because things like this
`CypressError: `cy.server()` was removed in Cypress version 12.0.0. Please update to use `cy.intercept()` instead.`

But thats not a problem, its just a function name change.

The rest are failing for things that no longer exist ej. Create Datasource fails because "Data source has been saved" is no longer displayed. (I'm testing this on a scada-lts 2.7.8 release too). I dont think that cypress tests are reliable in its current state.

Limraj replied:
@YuaFox,
1. You check release/2.8.0, which is the version we are currently working on.
2. 'But thats not a problem, its just a function name change.' What will you do about it?
3. 'Create Datasource fails because "Data source has been saved" is no longer displayed' Where? 
If it was displayed, it should still be like that, something must have gone broken...

YuaFox replied:
Hello @Limraj 
2. This commit fixes the deprecated function https://github.com/SCADA-LTS/Scada-LTS/pull/2990/commits/b87af6884c6cf04e72fca3f3d08ccd8b5a0f8b7a
3. I attach the cypress tests summary (I ran scadaLTS 2.7.8 to compare it with this pr)
[cypress_testing.txt](https://github.com/user-attachments/files/16815434/cypress_testing.txt) 
Yep, some things have gone broken, or changed but not reflected on tests.

I can fix the tests or fix features that are expected to work but I'll need some help :(

Limraj replied:
@YuaFox,
You need to check whether sometimes these tests did not break down on the old node. As for the operation of functions in the application, I think it is enough that the frontend compilation process runs successfully, then Node has nothing to do with the operation of the application itself. Am I wrong?

YuaFox replied:
Hello @Limraj 
Yes you're correct, Node has nothing to do with the operation of the application, scada-lts is only using it to compile the frontend, but i think cypress tests should be fixed and enforced to run correctly before merging. Maybe we should fix this in another task?

Limraj replied:
@YuaFox You just need to make sure whether there were problems with them on node 14 as well. If so, they can be corrected in a separate issue
.

Limraj replied:
Hi @YuaFox,
Can you confirm that the tests worked similarly on node 14? 
If so, I'll take a look at the code and we'll merge.
It is best to create a new issue to fix these tests.

Regards,
Kamil Jarmusik

Limraj replied:
Hi @YuaFox,
you still need to update the files:
- WebContent/resources/package-lock.json, 
- scadalts-ui/package-lock.json;

As follows:
1. We delete these two lock files;
2. Delete: 
- WebContent/resources/node_modules,
- scadalts-ui/node_modules (and rollback @min-gb);
3. We rebuild the project;

**I have these files updated, I can commit it to your branch if you don't mind.**

[#2805_Upgrade_Node__-_corrected__WebContent_resources_package-lock_json,_scadalts-ui_packa.patch](https://github.com/user-attachments/files/16950102/2805_Upgrade_Node__-_corrected__WebContent_resources_package-lock_json._scadalts-ui_packa.patch)

Regards.

YuaFox replied:
Hi @Limraj ,

If you have these files changed feel free to to commit them here.

The unit tests work as expected, the cypress tests there are some that worked on node 14 but not on node 22. I haven't had time to take a closer look and see if it's just a change in the dom or something similar or if they are real bugs.

Limraj replied:
Hi @YuaFox,
usage command:
`npx cypress run`
and Node version:
![image](https://github.com/user-attachments/assets/9437b679-ecf9-4b7f-ae6e-76429fbee41d)
then:
![image](https://github.com/user-attachments/assets/d1273c9f-188e-407d-b8d0-f22c49928a50)

Limraj commented on scadalts-ui/tests/utils/testing-utils.js:
```diff
@@ -82,6 +82,7 @@ export function prepareShallowMountWrapper(
 		vuetify,
 		store,
 		i18n,
+		router,
```
What is the justification for this?

YuaFox replied:
If the router is not added to tests, SynopticPanel testing don't work because this.$router is undefined.

https://github.com/SCADA-LTS/Scada-LTS/actions/runs/10505804935/job/29104184414

I dont get how testing was working before upgrading.

Limraj commented on scadalts-ui/.gitignore:
```diff
@@ -29,5 +29,7 @@ test-results.xml
+!@min-gb/dist
+!@min-gb/public
+!@min-gb/src
+!@min-gb/*.*
```
I copied min-gb and called min-gb2 and this pattern worked as we expected: (what has already been added to the git repository does not work with gitignore)
```
node_modules/*
!node_modules/@min-gb
node_modules/@min-gb/vuejs-components/*
!node_modules/@min-gb/vuejs-components/dist
!node_modules/@min-gb/vuejs-components/src
!node_modules/@min-gb/vuejs-components/public
!node_modules/@min-gb/vuejs-components/*.*
```
but:
```
node_modules/*
!@min-gb/dist
!@min-gb/public
!@min-gb/src
!@min-gb/*.*
```
not working for dir/file that have not been added to the git repository.

You can also add this to the main gitignore file (`\Scada-LTS\.gitignore`):
`WebContent/resources/js-ui/`

then only file change from repo:
![image](https://github.com/user-attachments/assets/f1007e56-ed72-4d07-b79d-13ce6753ef39)

YuaFox replied:
Sorry, I didn't notice there was a vuejs-components subfolder

Limraj replied:
Hi @YuaFox, 
You can also add this to the main gitignore file (\Scada-LTS\.gitignore):
WebContent/resources/js-ui/

YuaFox replied:
It's already there
![image](https://github.com/user-attachments/assets/3c6e287d-4df0-4c2e-a90b-103d603a8382)

Limraj replied:
@YuaFox,
Okay, but I guess this pattern doesn't work, see for yourself... maybe because at the beginning there is /

YuaFox replied:
Its working for me doing this command to force retracking of all files. 

```
git rm --cached -r . && git add .
```

After doing this i have 0 files tracked from WebContent/resources/js-ui/ folder, but i found a problem:
.gitignore says /tomcat/lib should not be tracked but its in the repo

Limraj replied:
Hi @YuaFox Ok, it actually works for me, it doesn't show these changes either. (/WebContent/resources/js-ui/)
However, tracking /tomcat/lib is enabled, if it is in the sources, there is not much other option... and I don't see a pattern that could exclude this folder from tracking, there is neither the word lib nor tomcat.

YuaFox committed:
#2805 Upgrade Node
- Removed deprecated cypress functions
- Removed modern_watch_list test

YuaFox committed:
#2805 Upgrade Node
- Fixed scadalts-ui .gitignore

Limraj committed:
#2805 Upgrade Node:
- update lock file: scadalts-ui/package-lock.json, WebContent/resources/package-lock.json

Limraj committed:
#2805 Upgrade Node:
- removed ^ form Cypress version, and update package-lock.json files;

Limraj committed:
#2805 Upgrade Node:
- corrected cypress config, added: fixturesFolder, specPattern;
- added ag-grid-community lib;

Limraj committed:
#2805 Upgrade Node:
- revert data_point_details.cy.js

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.0' into feature/#2805_Upgrade_Node

Limraj committed:
#2805 Upgrade Node:
- upgrade version Node to v22.9.0;

Limraj reviewed (approved):
There was a problem with the latest version 22.10.0 but rolling back to version 22.9.0 solved the problem. The Node version was set to a fixed version to exclude the possibility that the project would stop building for this reason. 

A separate issue has also been created for e2e test fixes.

https://github.com/SCADA-LTS/Scada-LTS/issues/3014
